commander.ts

```diff
@@ -228,11 +269,22 @@ export class Commander implements CmdVisitor<Value | null> {
     return tracer.spanSync('runCommand', () => {
       let chunk: Chunk;
       try {
-        chunk = compile(cmd, {
+        const result = compile(cmd, {
           filename: '<input>',
           cmdNo: this.cmdNo,
           cmdSource: this.cmdSource,
         });
+
+        chunk = result[0];
+
+        // TODO: The current behavior for interactive commands is to
+        // log the warnings as they occur. But what we should do is compile
+        // all interactive commands first, return that to evalCommands, then
+        // have that merge/log warnings before execution. This does mean
+        // that non-runtime commands will need to return a thunk.
```

Comment:
This doesn't necessarily have to be a thunk; that's just how you would avoid visiting twice. The alternative is to have a second visitor - either `Commander` returns the commands and an `Executor` actually runs the commands, or a `CommandCompiler` visitor returns `Chunk | Cmd` and `Commander` remains in charge of execution.

Of course, the vast majority of commands will be handled by the runtime, so I could also avoid a proper visitor and do a block of instanceof checks instead.

Reply:
In fact, the move might be to implement this in the compiler - split its interface up between commands and programs (similar to the parser), use a `IsInteractive` visitor to branch, and return an `Array<CompilerResult | Cmd>`.